### ISSUE TITLE
L2 caching information and optimization

### ISSUE THREAD

**Issue by onlinebizsoft** [magento/magento2#37421]
### Summary

We are a website with many store views and many products, our Magento redis cache is really big because how big our system is. I'm looking at L2 caching and hoping it will improve a bit our network transfer, network latency,......

However the document about L2 caching seems to be lacking technical information, it is not clear how it works because it is not possible to just store whole remote cache on the L2 cache and if the L2 cache is too small then even using L2 cache will just add extra load on the top (to save and remove L2 cache).

Can someone explain or share the experience with using this? 
CC @hostep @georgebabarus @Nuranto @jonathanribas @vzabaznov @Adel-Magebinary @igorwulff @drew7721 @ihor-sviziev

### Examples

L2 cache seems to just work as an extra cache level on local web server machine

### Proposed solution

We should be able to configure L2 cache to keep some certain cache keys or find a way to make it more effective, otherwise it will be almost useless for a website with big Redis cache size.

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by onlinebizsoft** [comment]
P/S : In my user case, we have some quite big cache keys because number of websites/store views and number of different modules so our cache size is very big and L2 seems to be not applicable however we plan to update the logic to keep only certain cache key in L2 caching and the rest will go to remote cache storage. This will help to reduce data transfer over network and reduce remote Redis engine CPU to serve big keys
Just sharing our plan, if any one has a better idea, please feel free to discuss.

**Comment by Nuranto** [comment]
Hi @onlinebizsoft 

I'm not sure to understand why you don't want to use fully L2 caching. What would be the benefit of using it only for a part of cache ? A network call is costly, for small entries too.

**Comment by onlinebizsoft** [comment]
@Nuranto 1. We can't use it fully because our total Redis size is very big and it is much bigger than the total web server memory so it makes no sense to replicate everything on L2 which lead to save/remove operation very often
2. I know the main thing is network call (even we have all components within a private network). However I'm thinking some very big keys will have much more effect than other small ones so at least I can eliminate big keys from network transfer and I will monitor the result from there.

**Comment by Nuranto** [comment]
Maybe the issue is that you have too much cache than necessary ? How much memory is redis using ?
Do you use remote storage ? There's a known bug that produce huge amount of unnecessary cache : https://github.com/magento/magento2/issues/35820

**Comment by onlinebizsoft** [comment]
@Nuranto you know a website with 100K products can be few GB of cache and the website has like 100 websites on the same installation so you can imagine how big it can be. And of course, we don't have any kind of web server with so much memory and memory should need to be spent for PHP processes instead of caching.

**Comment by ihor-sviziev** [comment]
I just wonder, maybe separating the cache for each website might help you?

**Comment by onlinebizsoft** [comment]
@ihor-sviziev yes actually it was an option I was thinking about. However it will make the infrastructure more complex (deployment process,....)  and it will lead to ineffective sharing resource, we always need to reserve enough resource for each group of websites. So I'm trying with simpler options first.

**Comment by Nuranto** [comment]
We have as much products, but only 3 websites indeed. I suppose that makes the difference. Also, we have servers nodes with 192Gb of RAM, which surely helps...
I think there's no other option left for you than rewrite (or write a new) L2 caching (see `RemoteSynchronizedCache` class)

**Comment by onlinebizsoft** [comment]
An off-topic question, do you use persistent database connection (and/or Redis persistent connection) @Nuranto @ihor-sviziev ?

**Comment by ihor-sviziev** [comment]
nope :(

**Comment by jonathanribas** [comment]
Hi @onlinebizsoft, sorry for the late reply, we did some tests yesterday enabling L2 cache on EFS (not `/dev/shm`).

We have very good results regarding network usage on Redis side, it has clearly dropped!

**Network usage without L2 cache:**

<img width="1778" alt="Screenshot 2023-09-13 at 10 08 07" src="https://github.com/magento/magento2/assets/3003782/74e69c8f-0511-4f97-9b87-8393f01efa16">

**Network usage with L2 cache enabled:**

<img width="1775" alt="Screenshot 2023-09-13 at 10 07 00" src="https://github.com/magento/magento2/assets/3003782/6f5efed0-5eb0-4eb7-95bf-b51db7b6e9f4">

**BUT we have very poor results in terms of PHP response consumption when enabling L2 cache:**

**PHP transaction time without L2 cache:**

<img width="1404" alt="Screenshot 2023-09-13 at 10 03 11" src="https://github.com/magento/magento2/assets/3003782/43da6a10-24d9-4cd0-bc5a-16b658606839">

**PHP transaction time with L2 cache:**

<img width="1402" alt="Screenshot 2023-09-13 at 10 03 41" src="https://github.com/magento/magento2/assets/3003782/ad83ee26-5e5e-42ed-9d00-124c4985a8bc">

**Comment by onlinebizsoft** [comment]
@jonathanribas it is a bit strange, currently we have tried to enable L2 on our development environment and we didn't notice such slowness. We are not ready to go live yet (our L2 implementation is a custom one to apply our logic for which key to be on L2 only)

**Comment by jonathanribas** [comment]
@onlinebizsoft, I've run this same test several times: without / with / without / with L2 cache and I got same results. Those environments are clean and without traffic.

**Comment by engcom-Hotel** [comment]
Hello @onlinebizsoft,

Thanks for the report and collaboration!

After going through the main description, seems like it can be a good feature to have, to configure the L2 cache to keep some certain cache keys or find a way to make it more effective.

Hence marking this issue as a feature request.

Thanks

**Comment by JamesFX2** [comment]
@jonathanribas why did you choose to use EFS rather than /dev/shm - this is surely built around use of in-memory store rather than disk reads of up to hundreds of identifiers? 

@onlinebizsoft did you ever follow up on this? To get what you'd want would probably involve storing last access times and purging based on that.

**Comment by onlinebizsoft** [comment]
> @jonathanribas why did you choose to use EFS rather than /dev/shm - this is surely built around use of in-memory store rather than disk reads of up to hundreds of identifiers?
> 
> @onlinebizsoft did you ever follow up on this? To get what you'd want would probably involve storing last access times and purging based on that.

I dont think it is a good approach because it will add too much logic and extra processing.

**Comment by JamesFX2** [comment]
@onlinebizsoft depends how you implement it, I agree adding a write to every read isn't the best but there are new approaches for storing in-memory cache. I am not an expert in this field but I've heard people talking enthusiastically about Swoole (in the context of explaining Laravel Octane).

**Comment by jonathanribas** [comment]
@JamesFX2, we run M2 on k8s and we don't want to change k8s default size setting for `/dev/shm`  which is `64mb`.

**Comment by onlinebizsoft** [comment]
@engcom-Hotel do we have any move on this? Im sure that the strongest point of Magento would be supporting complex system so optimizing for 100 stores system would be considered as high priority.

**Comment by Adel-Magebinary** [comment]
There might be a bug in paypal module. See here https://github.com/magento/magento2/issues/32118#issuecomment-1088545900

The issue was fixed after disabling the module.  CC: @onlinebizsoft

**Comment by Adel-Magebinary** [comment]
This is how you use L2 caching to save some network traffic. 

```
    'cache' => [
        'frontend' => [
            'default' => [
                'backend' => '\\Magento\\Framework\\Cache\\Backend\\RemoteSynchronizedCache',
                'backend_options' => [
                    'remote_backend' => '\\Magento\\Framework\\Cache\\Backend\\Redis',
                    'remote_backend_options' => [
                        'server' => 'redis-cluster-node-3',
                        'database' => '3',
                        'port' => '6379',
                        'compress_data' => '0',
                        'timeout' => '10'
                    ],
                    'local_backend' => 'Cm_Cache_Backend_File',
                    'local_backend_options' => [
                        'cache_dir' => '/dev/shm/'
                    ],
                    'use_stale_cache' => false
                ],
                'frontend_options' => [
                    'write_control' => false
                ]
            ],
            'type' => [
                'default' => [
                    'frontend' => 'default'
                ]
            ],
            'page_cache' => [
                'id_prefix' => 'ae1_',
                'backend' => 'Magento\\Framework\\Cache\\Backend\\Redis',
                'backend_options' => [
                    'server' => 'redis-cluster-node-3',
                    'database' => '2',
                    'port' => '6379',
                    'compress_data' => '0',
                    'timeout' => '10'
                ]
            ]
        ],

```
